base/secretbuffer.jl

```diff
@@ -0,0 +1,160 @@
+function SecretBuffer(str::String)
+    buf = codeunits(str)
+    s = SecretBuffer(sizehint=length(buf))
+    for c in buf
+        write(s, c)
+    end
+    seek(s, 0)
+    s
+end
+
+"""
+    SecretBuffer!(data::Vector{UInt8})
+
+Initialize a new `SecretBuffer` with `data` and securely zero the original source argument.
+"""
+function SecretBuffer!(d::Vector{UInt8})
+    len = length(d)
+    s = SecretBuffer(sizehint=len)
+    for i in 1:len
+        write(s, d[i])
+    end
+    seek(s, 0)
+    securezero!(d)
+    s
+end
+
+unsafe_SecretBuffer!(s::Cstring) = unsafe_SecretBuffer!(convert(Ptr{UInt8}, s), ccall(:strlen, Cint, (Cstring,), s))
+function unsafe_SecretBuffer!(p::Ptr{UInt8}, len=1)
+    s = SecretBuffer(sizehint=len)
+    for i in 1:len
+        write(s, unsafe_load(p, i))
+    end
+    seek(s, 0)
+    unsafe_securezero!(p, len)
+    s
+end
+
+
+show(io::IO, s::SecretBuffer) = print(io, "SecretBuffer(\"*******\")")
+
+hash(s::SecretBuffer, h::UInt) = hash(SecretBuffer, hash((s.data, s.size, s.ptr), h))
+==(s1::SecretBuffer, s2::SecretBuffer) = (view(s1.data, 1:s1.size), s1.ptr) == (view(s2.data, 1:s2.size), s2.ptr)
+
+function write(io::SecretBuffer, b::UInt8)
+    if io.ptr > length(io.data)
+        # We need to resize! the array: do this manually to ensure no copies are left behind
+        newdata = Vector{UInt8}(undef, (io.size+16)*2)
+        copyto!(newdata, io.data)
+        securezero!(io.data)
+        io.data = newdata
+    end
+    io.size == io.ptr-1 && (io.size += 1)
+    io.data[io.ptr] = b
+    io.ptr += 1
+    return 1
+end
+
+function write(io::IO, s::SecretBuffer)
+    nb = 0
+    for i in 1:s.size
+        nb += write(io, s.data[i])
+    end
+    return nb
+end
+
+cconvert(::Type{Cstring}, s::SecretBuffer) = unsafe_convert(Cstring, s)
+function unsafe_convert(::Type{Cstring}, s::SecretBuffer)
+    # Ensure that no nuls appear in the valid region
+    if any(==(0x00), s.data[i] for i in 1:s.size)
+        throw(ArgumentError("`SecretBuffers` containing nul bytes cannot be converted to C strings"))
+    end
+    # Add a hidden nul byte just past the end of the valid region
+    p = s.ptr
+    s.ptr = s.size + 1
+    write(s, '\0')
+    s.ptr = p
+    s.size -= 1
+    return Cstring(unsafe_convert(Ptr{Cchar}, s.data))
+end
+
+seek(io::SecretBuffer, n::Integer) = (io.ptr = max(min(n+1, io.size+1), 1); io)
+seekend(io::SecretBuffer) = seek(io, io.size+1)
+skip(io::SecretBuffer, n::Integer) = seek(io, position(io) + n)
+
+bytesavailable(io::SecretBuffer) = io.size - io.ptr + 1
+position(io::SecretBuffer) = io.ptr-1
+eof(io::SecretBuffer) = io.ptr > io.size
+isempty(io::SecretBuffer) = io.size == 0
+function peek(io::SecretBuffer)
+    eof(io) && throw(EOFError())
+    return io.data[io.ptr]
+end
+function read(io::SecretBuffer, ::Type{UInt8})
+    eof(io) && throw(EOFError())
+    byte = io.data[io.ptr]
+    io.ptr += 1
+    return byte
+end
+
+final_shred!(s::SecretBuffer) = shred!(s)
```

Comment:
Shouldn't it print a warning or something if `shred!` has not already been called?  Waiting until gc runs seems like a bad idea.